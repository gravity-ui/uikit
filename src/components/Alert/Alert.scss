@use '../variables';

$block: '.#{variables.$ns}alert';

#{$block} {
    &_corners_square {
        --g-card-border-radius: 0;
    }

    &__text-content {
        width: 100%;
    }

    &__actions_minContent {
        width: min-content;
    }
}
