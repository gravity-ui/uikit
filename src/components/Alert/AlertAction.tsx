'use client';

import React from 'react';

import {Button} from '../Button';

import type {AlertActionProps} from './types';
import {useAlertContext} from './useAlertContext';

export const AlertAction = (props: AlertActionProps) => {
    const {view} = useAlertContext();

    return <Button view={view === 'filled' ? 'normal-contrast' : undefined} {...props} />;
};
