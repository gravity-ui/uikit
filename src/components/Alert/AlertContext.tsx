import React from 'react';

import type {AlertContextType} from './types';

export const AlertContext = React.createContext<AlertContextType | null>(null);
