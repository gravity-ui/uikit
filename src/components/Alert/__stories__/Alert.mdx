import {Meta} from '@storybook/addon-docs';
import {
    DefaultUsage,
    Variants,
    WithPlaneActions,
    WithActionsOverride,
    WithCloseButton,
    WithHtmlMessageAndTitle,
    HorizontalLayoutAndCustomIcon,
} from './examples';
import * as LayoutStories from './Alert.stories.tsx';

<Meta of={LayoutStories} />

# Alert

User friendly communication by text

```tsx
import {Alert} from '@gravity-ui/uikit';

<Alert theme="info" title="Good news, everyone!" message="Today is a good day to do something!" />;
```

<DefaultUsage />

## Properties:

-   `title` - title of Alert component. Override if by `Alert.Title` component;
    ```tsx
    <Alert title={<Alert.Title className={'some-class'}>some text</Alert.Title>} />
    ```
-   `message` - body of Alert componet. One of `title` of `message` props are required;
-   `corners` - 'rounded' | 'square';
-   `layout` - override actions position (default bottom). Variants:

    -   vertical - bottom (by default);
    -   horizontal - right;

-   `view`:
    -   filled
    -   outlined
-   `theme`. By default `info`. [Variansts example](#variants):
    -   normal
    -   info
    -   positive
    -   warning
    -   danger
-   `hasClose`. Define function and close button will be appearing;
-   `actions`. Alert actions;
-   `icon` - slot to override default icon;
-   `align` - —Åenter your content in vertical direction. For example, you wanna icon component to be placed in center of card or height of your actions in horizontal direction more than content height;

### With plane actions:

```tsx
<Alert
    theme="danger"
    message="Are you sure?"
    actions={[
        {
            text: 'Do it!',
        },
        {
            text: 'No way!',
        },
    ]}
/>
```

<WithPlaneActions />

### With actions override and squere corners

```tsx
<Alert
    theme="positive"
    corners="square"
    title="Are you sure?"
    onClose={() => alert('Ok!')}
    view="outlined"
    actions={
        <Alert.Actions>
            <Button view="outlined">Do it!</Button>
            <Button view="flat">No way!</Button>
        </Alert.Actions>
    }
/>
```

<WithActionsOverride />

## With close button

```tsx
<Alert
    theme="info"
    title="Hey, Hero!"
    message="Wanna save the World?"
    onClose={() => alert('Ok!')}
    actions={<Button view="action">Do it!</Button>}
    // If you wont default aligment behaviour
    // actions={
    //     <Alert.Actions>
    //         <Button view="action">Do it!</Button>
    //     </Alert.Actions>
    // }
/>
```

<WithCloseButton />

## Html in message or title:

```tsx
<Alert
    theme="normal"
    title={<span dangerouslySetInnerHTML={{__html: '<b>Alert title</b>'}} />}
    message={<div dangerouslySetInnerHTML={{__html: '<i>Some message text</i>'}} />}
    onClose={() => alert('Ok!')}
/>
```

<WithHtmlMessageAndTitle />

## With horizontal layout and custom icon:

```tsx
<Alert
    theme="positive"
    align="center"
    icon={<Icon data={HandOk} size={16} />}
    title="Everything is ok, mister!"
    onClose={() => alert('Ok!')}
    layout="horizontal"
    actions={<Button view="outlined">Thanks</Button>}
/>
```

<HorizontalLayoutAndCustomIcon />

## Variants

<Variants />
