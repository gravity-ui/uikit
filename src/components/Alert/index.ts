export {Alert} from './Alert';
export type {AlertProps} from './types';
