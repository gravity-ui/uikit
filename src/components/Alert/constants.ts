import {block} from '../utils/cn';

import './Alert.scss';

export const DEFAULT_ICON_SIZE = 18;

export const bAlert = block('alert');
