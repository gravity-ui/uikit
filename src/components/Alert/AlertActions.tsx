'use client';

import React from 'react';

import {Flex} from '../layout';

import {AlertAction} from './AlertAction';
import {bAlert} from './constants';
import type {AlertActionsProps} from './types';
import {useAlertContext} from './useAlertContext';

export const AlertActions = ({items, children, className}: AlertActionsProps) => {
    const {layout} = useAlertContext();

    return (
        <Flex
            className={bAlert('actions', {minContent: layout === 'horizontal'}, className)}
            direction="row"
            gap="3"
            wrap
            alignItems={layout === 'horizontal' ? 'center' : 'flex-start'}
        >
            {items?.map(({handler, text}, i) => (
                <AlertAction key={i} onClick={handler}>
                    {text}
                </AlertAction>
            )) || children}
        </Flex>
    );
};
