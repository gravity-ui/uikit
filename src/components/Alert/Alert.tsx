import React from 'react';

import {Xmark} from '@gravity-ui/icons';

import {Button} from '../Button';
import {Card} from '../Card';
import {Icon} from '../Icon';
import {colorText} from '../Text';
import {Flex, spacing} from '../layout';

import {AlertAction} from './AlertAction';
import {AlertActions} from './AlertActions';
import {AlertContextProvider} from './AlertContextProvider';
import {AlertIcon} from './AlertIcon';
import {AlertTitle} from './AlertTitle';
import {DEFAULT_ICON_SIZE, bAlert} from './constants';
import type {AlertProps} from './types';

export const Alert = (props: AlertProps) => {
    const {
        theme = 'normal',
        view = 'filled',
        layout = 'vertical',
        message,
        className,
        corners,
        style,
        onClose,
        align,
        qa,
    } = props;

    return (
        <AlertContextProvider layout={layout} view={view}>
            <Card
                style={style}
                className={bAlert({corners}, spacing({py: 4, px: 5}, className))}
                theme={theme}
                view={view}
                qa={qa}
            >
                <Flex gap="3" alignItems={align}>
                    {typeof props.icon === 'undefined' ? (
                        <Alert.Icon theme={theme} view={view} />
                    ) : (
                        props.icon // ability to pass `null` as `icon` prop value
                    )}
                    <Flex direction={layout === 'vertical' ? 'column' : 'row'} gap="5" grow>
                        <Flex gap="2" grow className={bAlert('text-content')}>
                            <Flex direction="column" gap="1" grow justifyContent={align}>
                                {typeof props.title === 'string' ? (
                                    <Alert.Title text={props.title} />
                                ) : (
                                    props.title
                                )}
                                {message}
                            </Flex>
                        </Flex>
                        {Array.isArray(props.actions) ? (
                            <Alert.Actions items={props.actions} />
                        ) : (
                            props.actions
                        )}
                    </Flex>
                    {onClose && (
                        <Button view="flat" onClick={onClose}>
                            <Icon
                                data={Xmark}
                                size={DEFAULT_ICON_SIZE}
                                className={colorText({color: 'secondary'})}
                            />
                        </Button>
                    )}
                </Flex>
            </Card>
        </AlertContextProvider>
    );
};

Alert.Icon = AlertIcon;
Alert.Title = AlertTitle;
Alert.Actions = AlertActions;
Alert.Action = AlertAction;
