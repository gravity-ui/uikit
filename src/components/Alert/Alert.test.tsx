import React from 'react';

import userEvent from '@testing-library/user-event';

import {render, screen} from '../../../test-utils/utils';
import {Flex} from '../layout';

import {Alert} from './Alert';
import {bAlert} from './constants';
import type {AlertTheme} from './types';

const title = 'Where will you go, hero?';
const message = 'Choose wisely: the end of the fairy tale depends on your decision';
const right = 'To the right (lose the horse)';
const center = 'Straight (find a wife)';
const left = 'To the left (CC 235.2)';

describe('Alert', () => {
    test('render close button if callback passed', async () => {
        const callback = jest.fn();

        render(<Alert title={title} message={message} onClose={callback} />);

        const buttons = await screen.findAllByRole('button');

        expect(buttons.length).toEqual(1);

        await userEvent.click(buttons[0]);

        expect(callback).toBeCalledTimes(1);
    });

    test('render actions as buttons if decl passed', async () => {
        render(
            <Alert
                title={title}
                message={message}
                actions={[
                    {
                        text: right,
                    },
                    {
                        text: center,
                    },
                    {
                        text: left,
                    },
                ]}
            />,
        );

        expect((await screen.findAllByRole('button')).length).toEqual(3);
    });

    test('render actions as is if component passed', async () => {
        render(<Alert title={title} message={message} actions={<Flex>{center}</Flex>} />);

        expect(await screen.findByText(center)).toBeInTheDocument();
    });

    test('has predicted styles if inline layout rendered', async () => {
        const {container} = render(
            <Alert
                theme="danger"
                align="center"
                title={title}
                layout="horizontal"
                onClose={jest.fn()}
                actions={[{text: right}]}
            />,
        );

        expect(container).toMatchSnapshot();
    });

    test.each<AlertTheme>(['danger', 'info', 'success', 'warning', 'utility'])(
        'render correct icon if not normal theme',
        async (theme) => {
            const {container} = render(
                <Alert theme={theme} title={title} message={message} onClose={jest.fn()} />,
            );

            // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
            const icon = await container.querySelector(`.${bAlert('icon')}`);

            expect(icon).toBeInTheDocument();
        },
    );

    test("by default don't render icon component", async () => {
        const {container} = render(<Alert title={title} message={message} onClose={jest.fn()} />);

        // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
        const icon = await container.querySelector(`.${bAlert('icon')}`);

        expect(icon).not.toBeInTheDocument();
    });

    test('can render custom icon component if needed', async () => {
        const customIconText = "i'ma icon";
        render(<Alert title={title} message={message} icon={<div>{customIconText}</div>} />);

        expect(await screen.findByText(customIconText)).toBeInTheDocument();
    });
});
