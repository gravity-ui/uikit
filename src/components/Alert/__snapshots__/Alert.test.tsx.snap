// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert has predicted styles if inline layout rendered 1`] = `
<div>
  <div
    class="g-box g-card g-card_theme_danger g-card_view_filled g-card_type_container g-card_size_m g-alert g-s__py_4 g-s__px_5"
  >
    <div
      class="g-box g-flex"
      style="column-gap: 12px; row-gap: 12px; align-items: center;"
    >
      <div
        class="g-alert__icon g-color-text g-color-text_color_danger"
      >
        <svg
          aria-hidden="true"
          class="g-icon"
          fill="currentColor"
          height="18"
          stroke="none"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
        >
          <svg
            fill="none"
            viewBox="0 0 16 16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8 15A7 7 0 1 0 8 1a7 7 0 0 0 0 14ZM6.53 5.47a.75.75 0 0 0-1.06 1.06L6.94 8 5.47 9.47a.75.75 0 1 0 1.06 1.06L8 9.06l1.47 1.47a.75.75 0 1 0 1.06-1.06L9.06 8l1.47-1.47a.75.75 0 1 0-1.06-1.06L8 6.94 6.53 5.47Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </svg>
      </div>
      <div
        class="g-box g-flex"
        style="flex-direction: row; flex-grow: 1; column-gap: 20px; row-gap: 20px;"
      >
        <div
          class="g-box g-flex g-alert__text-content"
          style="flex-grow: 1; column-gap: 8px; row-gap: 8px;"
        >
          <div
            class="g-box g-flex"
            style="flex-direction: column; flex-grow: 1; column-gap: 4px; row-gap: 4px; justify-content: center;"
          >
            <span
              class="g-text g-text_variant_subheader-2 g-alert__title"
            >
              Where will you go, hero?
            </span>
          </div>
        </div>
        <div
          class="g-box g-flex g-alert__actions g-alert__actions_minContent"
          style="flex-direction: row; flex-wrap: wrap; column-gap: 12px; row-gap: 12px; align-items: center;"
        >
          <button
            class="g-button g-button_view_normal-contrast g-button_size_m g-button_pin_round-round"
            type="button"
          >
            <span
              class="g-button__text"
            >
              To the right (lose the horse)
            </span>
          </button>
        </div>
      </div>
      <button
        aria-label="Close"
        class="g-button g-button_view_flat g-button_size_m g-button_pin_round-round"
        type="button"
      >
        <span
          class="g-button__icon"
        >
          <span
            class="g-button__icon-inner"
          >
            <svg
              aria-hidden="true"
              class="g-icon g-color-text g-color-text_color_secondary"
              fill="currentColor"
              height="18"
              stroke="none"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
              xmlns:xlink="http://www.w3.org/1999/xlink"
            >
              <svg
                fill="none"
                viewBox="0 0 16 16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M3.47 3.47a.75.75 0 0 1 1.06 0L8 6.94l3.47-3.47a.75.75 0 1 1 1.06 1.06L9.06 8l3.47 3.47a.75.75 0 1 1-1.06 1.06L8 9.06l-3.47 3.47a.75.75 0 0 1-1.06-1.06L6.94 8 3.47 4.53a.75.75 0 0 1 0-1.06Z"
                  fill="currentColor"
                  fill-rule="evenodd"
                />
              </svg>
            </svg>
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
